OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs

```diff
@@ -0,0 +1,171 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class GrantConditionOnAttackInfo : ITraitInfo
+	{
+		[GrantedConditionReference]
+		[Desc("The condition type to grant.")]
+		public readonly string Condition = null;
+
+		[Desc("Name of the armaments that grant this condition.")]
+		public readonly HashSet<string> ArmamentNames = new HashSet<string>();
+
+		[Desc("Shots required to apply an instance of the condition. If there are more instances of the condition granted than values listed,",
+			"the last value is used for all following instances beyond the defined range.")]
+		public readonly int[] RequiredShotsPerInstance = { 1 };
+
+		[Desc("Maximum instances of the condition to grant.")]
+		public readonly int MaximumInstances = 1;
+
+		[Desc("Should all instances reset if the actor passes the final stage?")]
+		public readonly bool IsCyclic = false;
+
+		[Desc("Amount of ticks required to pass without firing to revoke an instance.")]
+		public readonly int RevokeDelay = 15;
+
+		[Desc("Should an instance be revoked if the actor changes target?")]
+		public readonly bool RevokeOnNewTarget = false;
+
+		[Desc("Should all instances be revoked instead of only one?")]
+		public readonly bool RevokeAll = false;
+
+		public object Create(ActorInitializer init) { return new GrantConditionOnAttack(init, this); }
+	}
+
+	public class GrantConditionOnAttack : INotifyCreated, ITick, INotifyAttack
+	{
+		readonly GrantConditionOnAttackInfo info;
+		readonly Stack<int> tokens = new Stack<int>();
+
+		int cooldown = 0;
+		int shotsFired = 0;
+		ConditionManager manager;
+		Target lastTarget = Target.Invalid; // only tracked when RevokeOnNewTarget is true.
+
+		public GrantConditionOnAttack(ActorInitializer init, GrantConditionOnAttackInfo info)
+		{
+			this.info = info;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			manager = self.Trait<ConditionManager>();
```

Comment:
`TraitOrDefault` otherwise the null checks are pointless or this is missing a `Requires<ConditionManagerInfo>`.

Reply:
I added Requires<ConditionManagerInfo> as this module is completely useless without the condition manager.

Reply:
As mentioned above, that's not a great option.  Can you please make this `TraitOrDefault` and put the null checks back in so this matches the behaviour of the other traits?